src/invariant/EventsAreConsistentWithEntryDiffs.cpp

```diff
@@ -87,62 +117,78 @@ getAmountFromData(SCVal const& data)
     return I128ZERO;
 }
 
-CxxI128
-consumeAmount(
-    SCAddress const& addr, Asset const& asset,
-    UnorderedMap<SCAddress, UnorderedMap<Asset, CxxI128>>& eventAmounts)
+std::optional<SCAddress>
+getAddressFromBalanceKey(LedgerKey const& lk)
 {
-    auto lkAssetMapIt = eventAmounts.find(addr);
-    if (lkAssetMapIt == eventAmounts.end())
+    if (lk.contractData().key.type() != SCV_VEC)
     {
-        return I128ZERO;
+        return std::nullopt;
     }
 
-    auto& lkAssetMap = lkAssetMapIt->second;
-    auto assetAmountIt = lkAssetMap.find(asset);
-    if (assetAmountIt == lkAssetMap.end())
+    auto const& vec = lk.contractData().key.vec();
+    if (!vec || vec->size() != 2)
     {
-        return I128ZERO;
+        return std::nullopt;
     }
 
-    auto res = assetAmountIt->second;
+    auto const& addr = vec->at(1);
 
-    // Now remove this value from the map
-    lkAssetMap.erase(assetAmountIt);
-    if (lkAssetMap.empty())
+    if (addr.type() != SCV_ADDRESS)
     {
-        eventAmounts.erase(lkAssetMapIt);
+        return std::nullopt;
     }
-    return res;
+
+    return addr.address();
 }
 
-std::optional<SCAddress>
-getAddressFromBalanceKey(LedgerKey const& lk)
+// Should only be called with trustlines as input
+bool
+checkAuthorization(AggregatedEvents& agg, SCAddress const& trustlineOwner,
+                   Asset const& asset, LedgerEntry const* current,
+                   LedgerEntry const* previous)
 {
-    if (lk.contractData().key.type() != SCV_VEC)
+    if (!current || !previous)
     {
-        return std::nullopt;
+        // If a trustline was created or deleted auth could not have changed.
+        return true;
     }
 
-    auto const& vec = lk.contractData().key.vec();
-    if (!vec || vec->size() != 2)
+    bool currAuth = isAuthorized(*current);
+    bool prevAuth = isAuthorized(*previous);
+
+    std::optional<bool> eventAuth;
+
+    auto assetMapIt = agg.mIsAuthorized.find(trustlineOwner);
+    if (assetMapIt != agg.mIsAuthorized.end())
     {
-        return std::nullopt;
+        auto authIt = assetMapIt->second.find(asset);
+        if (authIt != assetMapIt->second.end())
+        {
+            eventAuth = authIt->second;
+        }
     }
 
-    auto const& addr = vec->at(1);
+    auto res = true;
 
-    if (addr.type() != SCV_ADDRESS)
+    // if event was seen, make sure it matches current state
+    if (eventAuth)
     {
-        return std::nullopt;
+        res = (*eventAuth == currAuth);
     }
 
-    return addr.address();
+    // if auth on trustline changed, make sure an event was emitted
+    if (res && currAuth != prevAuth)
```

Comment:
This check seems a unnecessarily complicated and it has some duplication. IIUC what you want to check is
```
// Auth changed => emit correct event
if (currAuth != prevAuth) { 
   return (eventAuth && (*eventAuth == currAuth)); 
} else { 
  // Auth has not changed => no event emitted.
  return !eventAuth; 
}
```

Reply:
I see. That's a bit surprising. Still, I think duplication is not necessary - wouldn't it be sufficient to verify that event is present in case if auth has changed? (we've already made sure that the event corresponds to the correct end state of the trustline).

Reply:
Ah yeah I see what you're saying. Updated the code.